metacheck/main.c

```diff
@@ -0,0 +1,406 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Meta Check
+ *
+ * Metadata checker for imx6ull NAND flash
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Maciej Purski
+ *
+ */
+
+/*
+* This program is free software; you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation; either version 2 of the License, or
+* (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License along
+* with this program; if not, write to the Free Software Foundation, Inc.,
+* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include <errno.h>
+#include <limits.h>
+#include <stdio.h>
+#include <imx6ull-flashsrv.h>
+
+#include <sys/mman.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <sys/types.h>
+#include <sys/msg.h>
+#include <sys/stat.h>
+
+#include "bch.h"
+
+#define PAGE_SZ        (4096 + 256)
+#define META_SZ        16
+#define METAECC_SZ     26
+#define CLEANMARKER_SZ (META_SZ + METAECC_SZ)
+
+#define METAECC_STRENGTH 16
+#define METAECC_GF       13
+
+
+typedef struct partition {
+	oid_t oid;
+	uint32_t rawpagesz;
+	uint32_t nblocks;
+	uint32_t rawblocksz;
+} partition_t;
+
+
+static struct {
+	uint8_t buf[PAGE_SZ];
+	partition_t *partitions;
+	int nparts;
+	int pageMode;
+	int verbose;
+} common;
+
+
+static int readraw(oid_t oid, uint32_t addr, void *data, int size)
+{
+	msg_t msg = { 0 };
+	flash_i_devctl_t *idevctl = (flash_i_devctl_t *)msg.i.raw;
+	flash_o_devctl_t *odevctl = (flash_o_devctl_t *)msg.o.raw;
+
+	msg.type = mtDevCtl;
+	msg.o.data = data;
+	msg.o.size = size;
+
+	idevctl->type = flashsrv_devctl_readraw;
+	idevctl->readraw.oid = oid;
+	idevctl->readraw.size = size;
+	idevctl->readraw.address = addr;
+
+	if (msgSend(oid.port, &msg) < 0) {
+		return -1;
+	}
+
+	if (odevctl->err != size) {
+		return odevctl->err;
+	}
+
+	return 0;
+}
+
+
+static int containsCleanmarker(const uint8_t *buf)
+{
+	static const uint8_t cleanmarker[] = { 133, 25, 3, 32, 8, 0, 0, 0, 255, 255, 255, 255, 255, 255, 255, 255 };
+
+	return !memcmp(buf, cleanmarker, sizeof(cleanmarker));
+}
+
+
+static int isBadBlock(const uint8_t *buf)
+{
+	if (buf[0] == 0 && buf[1] == 0) {
+		return 1;
+	}
+
+	return 0;
```

Comment:
If we don't care about MISRA (I can't really tell from the code :D )
```suggestion
	return !buf[0] && !buf[1];
```

Reply:
That's too much :D I like my version more in this case.